File: 217. Contains Duplicate/sol.cpp
#include <bits/stdc++.h>
using namespace std;
// T.C = O(n)
// S.c = O(n)
class Solution {
public:
    bool containsDuplicate(vector<int>& nums) {
        map<int,int>m;
        for(int i = 0; i < nums.size(); i++){
            if(m.find(nums[i]) != m.end()){
                return true;
            }else{
                m[nums[i]] = 1;
            }
        }
        return false;
    }
};
int main(){
    vector<int> nums = {1,2,3,4};
    Solution s;
    cout << s.containsDuplicate(nums) << endl;
    return 0;
}

File: 153. Find Minimun in Rotated Sorted Array/sol.py
#case 2: rotate 1 .. n -1, min in the mid of list

#case 1: rotate n times, min in left of list

class Solution:
    def findMin(self, nums: List[int]) -> int:
        n = len(nums)
        left = 0
        right = n - 1
        mid = (n - 1)//2

        if(nums[right] > nums[left]):
            return nums[left]      # case 1
        # case 2: find the min in the mid of list
        while left < right:
            mid = (right + left) // 2
            if nums[mid] > nums[right]:
                left = mid + 1
            else:
                right = mid
        return nums[left]
            
